### DISCUSSION TITLE
Nested forms

### DISCUSSION THREAD

**Discussion by gadiGuesty** [react-hook-form/react-hook-form#7038] [Q&A]
Hey all, what is best practice when this kind of ux is given: [Nedted form example](https://codesandbox.io/embed/p7p6lx24xq?codemirror=1). HTML 5 forbid form nesting:[https://stackoverflow.com/questions/379610/can-you-nest-html-forms](https://stackoverflow.com/questions/379610/can-you-nest-html-forms). though seems like there are hacks around it even with react hook form:[Nested form in React hook form](https://gist.github.com/Gallevy3313/5d1241f359867de058bd3b58a85c1b70). What would be the right way to do it?

**Comment by JamesWong1999** [comment]
in react hook form, you probably want to build multiple forms. so you can have a main form and update the main form with sub form values.

**Reply by yleflour** [reply]
You can use portals: https://codesandbox.io/s/react-hook-form-nested-portal-6x3fvy?file=/src/App.tsx

**Reply by theshubhagrwl** [reply]
As far as I understand, the work of portal is to render the form. For me the rendering was not an issue and hence I didn't use portals for it. I simply followed [this](https://www.react-hook-form.com/advanced-usage/#ConnectForm) for rendering out the form and it works. The real issue was I wanted to submit the nested form first and then the main form. 

Correct me if I am wrong here.

**Reply by metawort** [reply]
I tried using portals, but even though the inner form in the portal is correctly rendered outside the outer form, submitting the inner form still also submits the outer form. I can fix this with `stopPropagation()` on the inner submit event, but I'm still baffled why this is happening...

**Reply by yleflour** [reply]
@metawort As stated in my example, events go through React's DOM instead of HTML's DOM. For react, your form is still considered as nested.

**Comment by akucharczyk** [comment]
Just use in the form `event.stopPropagation()`

To call `handleSubmit` probably just call it like this: `handleSubmit(yourSubmitFunction)(event)`

**Reply by M4Marvin** [reply]
But the event is deprecated.

**Reply by mshahzebraza** [reply]
this is not scalable either.
Suppose you're building a multi-step form, with say 5 stages.

each of these stages might contain a nested dialog form.

You wouldn't want to go for each of these nested forms and stop propagation from there.
I faced a similar issue earlier, and in the parent form, I just checked if the submission is coming from the allowed targets (by equating the target.name with the allowed target) and then assigning the name to the allowed submission trigger.

That way I just modify the parent form handler once, and in future if I add multiple nested forms, I don't need to worry about it

**Comment by tatsmaki** [comment]
You should definitely look into the portals. Nested form is not a valid HTML structure – this can impact both UX and SEO.
The portal can render your modal window in the outer DOM, so your form will no longer be nested.

**Reply by mateuszJS** [reply]
Good solution! Unfortunately won't work with react-hook-form and `<FormProvider>`, on triggering submit event all submit callbacks are triggered which are nested in React DOM (not in the actually DOM in browser). So you final HTML output might be a valid HTML, yet issue persists. Can be solved via checking `e.target !== e.currentTarget` on submit event of the form that nests other forms inside. Or `preventDefault` and `stopPropagation` can be called on nested forms while submit.

**Comment by benmarg** [comment]
Okay finally figured out the answer to this, this issue has been haunting me for way too long: 

```ts
 <form
     onSubmit={form.handleSubmit(onSubmit)}
     id="unit-skill-dialog"
  >
```
```ts
<Button type="submit" form="unit-skill-dialog">
  Submit
</Button>
```

Add the id and form attributes to the `<form>` and `<submit>` button on the *inner* RHF instance, and this will prevent the event from propagating to the parent, no need to use `e.stopPropogation`

**Reply by hmaesta** [reply]
Hey, Benjamin!  [I created a Codesandbox](https://codesandbox.io/p/sandbox/muddy-glitter-cmj6yc) to reproduce your suggestion, but it didn't work. Is there anything I'm missing?

**Reply by mshahzebraza** [reply]
@hmaesta You're right. This doesn't work. THe nested forms stil submit.

This solution worked: for me https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-11228568

**Comment by mwarger** [comment]
The easiest way I've seen to get this working (as none of the tricks about `form` have worked for me) is to just not use `form` at all and trigger the submit from the button itself.

Advantages are not having to worry about form IDs (which did not work in my app - I don't know why), not having to worry about invalid form nesting, not having to worry about which form events to call or not.  Doing it this way, I can have separate sub forms each handled by RHF with different sub-schemas as re-usable pieces I can use throughout my app, but they communicate a small set of values up to their parent form with a callback that is called whenever the subform is done with its successful submission.

Something like this:

```
<button
  type="button"
  onClick={handleSubmit(onSubmit, onInvalid)}
>
  Create
</button>
```

**Reply by Aklile-Yilma** [reply]
It works for me

**Reply by Steveb599** [reply]
where does the onInvalid comes from?

**Reply by BrendanC23** [reply]
That's a custom function you write to handle errors when the form is submitted (such as an input not being filled out). See the [handleSubmit docs](https://react-hook-form.com/docs/useform/handlesubmit).

```tsx
function onSubmit(data) {
    // ...
}

function onInvalid(errors) {
    // ...
}

return (
<button
  type="button"
  onClick={handleSubmit(onSubmit, onInvalid)}
>
  Submit
</button>
)
```

**Comment by dwilt** [comment]
I really struggled to get this work for a while and there are [pieces](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-3190034) of [correct](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-7553521) information in this thread but IMO, nothing that pulls it all together and gives a valid HTML markup and accessible solution. Going to try and solve that here:

If you have nested `<form>` elements _in the DOM_, you've got invalid HTML and you should start by fixing this. Regardless of RHF, you could run into accessibility/usability/weird behavior issues with this and I really advise against it. So, if you do have a nested `<form>`s,  there's a good chance you can pull the inner forms out from the parent `<form>` with a [portal](https://react.dev/reference/react-dom/createPortal).

**Now**, you need to also solve the context issue (nested `<FormProvider>` elements) which is also at play here (which others have talked about).  When you try to use `handleSubmit`, all of the `handleSubmit`s in context get called. You can solve this without needing to get rid of the the nested `<FormProvider>`s and [by utilizing `e.stopPropagation()` and `e.preventDefault()`](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-7553521):

```JSX
<form
  onSubmit={(e) => {
    e.stopPropagation();
    e.preventDefault();

    // the void is for eslint because `handleSubmit` returns a promise.
    void handleSubmit(({ dataBinding }) => {
      onSave(dataBinding);
    })(e);
  }}
>
 ```
 
- There is no need for an [`id` on the `form` or the `form`](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-10782081) on the submit button.
-  You should really always use `<form>`s whenever possible because they're important for accessibility and having valid markup so I although [@mwarger's suggestion](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-10789928) works, it's not a valid way of keeping your markup.
- You can keep your nested `<FormProvider>` elements
- If you need to [support a multi-step form](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-11228568), this solution should also work as well

**Reply by gauravnepal3** [reply]
Works for me, Thanks

**Reply by yleflour** [reply]
I forgot to mention it in this thread but I published a nested form example 3 years ago, available in the official docs: https://github.com/react-hook-form/react-hook-form/tree/master/examples.

The example link is https://codesandbox.io/s/react-hook-form-nested-portal-bw8m75

If this example doesn't work for you (@dwilt), I believe react-hook-form's team would welcome a new one with a pull request to update the docs

**Reply by mateuszJS** [reply]
Another option can be having a guard on the "top" form which holds nested forms inside.
```
<form
  onSubmit={(e) => {
    if (e.target !== e.currentTarget) return;

    void handleSubmit(({ dataBinding }) => {
      onSave(dataBinding);
    })(e);
  }}
>
```
In my case I have multiple nested forms in a dialogs(portals), so updating just one form seems like a simpler solution.
